### PR TITLE
Document terraform stack deployment modules

### PR DESCRIPTION
**Pull request by stefannica**
## Describe changes
Document [our newly implemented Terraform stack deployment modules](https://registry.terraform.io/modules/zenml-io/zenml-stack).

Sync-ed with: https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/sSi8uXbe2nPYIyiHcpxS/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Other (add details above)

### CODE REVIEW

File: docs/book/component-guide/artifact-stores/azure.md [hunk 1]
```diff
@@ -22,10 +22,26 @@ You should use the Azure Artifact Store when you decide to keep your ZenML artif
...
-or the [easy cloud registration wizard](../../how-to/stack-deployment/register-a-cloud-stack.md)
-for a shortcut on how to deploy & register this stack component.
+Would you like to skip ahead and deploy a full ZenML cloud stack already,
+including an Azure Artifact Store?
+
+You can use one of several different shortcuts that ZenML provides you to help
+you deploy your first stack based on real cloud infrastructure with minimal
+effort:
+
+* you can let ZenML at the wheel and use [the automated 1-click stack deployment flow](../../how-to/stack-deployment/deploy-a-cloud-stack.md) to build your ZenML cloud stack from scratch. This is the easiest way to get started with cloud ZenML stacks, in the dashboard or with the CLI:
```
Comment (strickvl) [inline docs/book/component-guide/artifact-stores/azure.md:32]:
```suggestion
* you can let ZenML take the wheel and use [the automated 1-click stack deployment flow](../../how-to/stack-deployment/deploy-a-cloud-stack.md) to build your ZenML cloud stack from scratch. This is the easiest way to get started with cloud ZenML stacks, in the dashboard or with the CLI:
```

File: docs/book/component-guide/artifact-stores/gcp.md [hunk 2]
```diff
@@ -22,12 +22,29 @@ You should use the GCS Artifact Store when you decide to keep your ZenML artifac
...
-or the [easy cloud registration wizard](../../how-to/stack-deployment/register-a-cloud-stack.md)
-for a shortcut on how to deploy & register this stack component.
+Would you like to skip ahead and deploy a full ZenML cloud stack already,
+including a GCS Artifact Store?
+
+You can use one of several different shortcuts that ZenML provides you to help
+you deploy your first stack based on real cloud infrastructure with minimal
+effort:
+
+* you can let ZenML at the wheel and use [the automated 1-click stack deployment flow](../../how-to/stack-deployment/deploy-a-cloud-stack.md) to build your ZenML cloud stack from scratch. This is the easiest way to get started with cloud ZenML stacks, in the dashboard or with the CLI:
```
Comment (strickvl) [inline docs/book/component-guide/artifact-stores/gcp.md:32]:
```suggestion
* you can let ZenML take the wheel and use [the automated 1-click stack deployment flow](../../how-to/stack-deployment/deploy-a-cloud-stack.md) to build your ZenML cloud stack from scratch. This is the easiest way to get started with cloud ZenML stacks, in the dashboard or with the CLI:
```

File: docs/book/component-guide/artifact-stores/gcp.md [hunk 3]
```diff
@@ -22,12 +22,29 @@ You should use the GCS Artifact Store when you decide to keep your ZenML artifac
...
+zenml stack deploy --provider gcp
+```
+* if you already have some infrastructure pieces provisioned in your cloud and you prefer to have more control over the stack configuration, 
+you can also use [the stack wizard](register-a-cloud-stack.md) to seamlessly register your ZenML cloud stack in the dashboard or with the CLI:
+
+```shell
+zenml stack register <STACK_NAME> --provider gcp
+```
+
+* if you are a hardened infrastructure professional who prefers a more hands-on approach that gives you more control over the infrastructure that is provisioned, you can use [our GCP Terraform modules](../../how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md) to deploy a ZenML cloud stack with Terraform.
```
Comment (strickvl) [inline docs/book/component-guide/artifact-stores/gcp.md:44]: 💪

File: docs/book/component-guide/artifact-stores/s3.md [hunk 4]
```diff
@@ -22,10 +22,26 @@ You should use the S3 Artifact Store when you decide to keep your ZenML artifact
...
-or the [easy cloud registration wizard](../../how-to/stack-deployment/register-a-cloud-stack.md)
-for a shortcut on how to deploy & register this stack component.
+Would you like to skip ahead and deploy a full ZenML cloud stack already,
+including an S3 Artifact Store?
+
+You can use one of several different shortcuts that ZenML provides you to help
+you deploy your first stack based on real cloud infrastructure with minimal
+effort:
+
+* you can let ZenML at the wheel and use [the automated 1-click stack deployment flow](../../how-to/stack-deployment/deploy-a-cloud-stack.md) to build your ZenML cloud stack from scratch. This is the easiest way to get started with cloud ZenML stacks, in the dashboard or with the CLI:
```
Comment (strickvl) [inline docs/book/component-guide/artifact-stores/s3.md:32]:
```suggestion
* you can let ZenML take the wheel and use [the automated 1-click stack deployment flow](../../how-to/stack-deployment/deploy-a-cloud-stack.md) to build your ZenML cloud stack from scratch. This is the easiest way to get started with cloud ZenML stacks, in the dashboard or with the CLI:
```

(and anywhere else this phrase is to be found...)

File: docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+
+ZenML maintains a collection of [Terraform modules](https://registry.terraform.io/modules/zenml-io/zenml-stack)
+designed to streamline the provisioning of cloud resources and seamlessly
+integrate them with ZenML Stacks. These modules simplify the setup process,
+allowing users to quickly provision cloud resources as well as configure and
+authorize ZenML to utilize them for running pipelines and other AI/ML operations.
+
+By leveraging these Terraform modules, users can ensure a more efficient and
+scalable deployment of their machine learning infrastructure, ultimately
+enhancing their development and operational workflows. The modules
```
Comment (strickvl) [inline docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md:15]:
```suggestion
enhancing their development and operational workflows. The modules'
```

File: docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+other things, this means that you will need to have Terraform installed on your
+machine and you will need to manually manage the state of your infrastructure.
+
+If you prefer a more automated approach, you can use [the 1-click stack deployment feature](deploy-a-cloud-stack.md)
+to deploy a cloud stack with ZenML with minimal knowledge of Terraform or cloud
+infrastructure for that matter.
+
+If you have the required infrastructure pieces already deployed on your cloud, 
+you can also use [the stack wizard to seamlessly register your stack](register-a-cloud-stack.md).
+{% endhint %}
```
Comment (strickvl) [inline docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md:30]:
```suggestion
you can also use [the stack wizard to seamlessly register your stack](register-a-cloud-stack.md).
{% endhint %}
```

I'm not sure that these Markdown links show up as links when inside these hints... I think you'll have to use an HTML <a> tag instead, but test it out in gitbook preview to confirm..

**Review by strickvl — CHANGES_REQUESTED** [review]

File: docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md (hunk 6 above)
Comment (stefannica) [reply docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md]: you were right about this. I need to use relative links. Absolute links don't work here.

Comment (stefannica) [reply docs/book/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform.md]: updated and verified that all links work now

**Review by htahir1 — CHANGES_REQUESTED** [review]
So in general i like this but I feel like there is too much information sort of "dumped" here. The art of writing docs is also to make it consumable and easily digestible for the user. While this is super structured chunks like this can be very daunting for users: 

image

This is when somebody is doing the production guide and it is repeated many places where I feel it comes as a jarring 3-option choice. Before the callout was a single sentence. We should keep it that way, and add the terraform option 

That's my feedback :-) Let me know if you'd like to be more specific but I generally think a fresh look at this with the lens of a user who as limited time would help here!

**Comment by stefannica** [comment]
> So in general i like this but I feel like there is too much information sort of "dumped" here. The art of writing docs is also to make it consumable and easily digestible for the user. While this is super structured chunks like this can be very daunting for users:
> 
> image
> This is when somebody is doing the production guide and it is repeated many places where I feel it comes as a jarring 3-option choice. Before the callout was a single sentence. We should keep it that way, and add the terraform option
> 
> That's my feedback :-) Let me know if you'd like to be more specific but I generally think a fresh look at this with the lens of a user who as limited time would help here!

Alright, fair point. How's this ?

image

**Comment by htahir1** [comment]
@stefannica id rather say replace "easy cloud deployment wizard" and "easy cloud registration wizard" with "in-browser stack deployment wizard" and "stack registration wizard"

**Review by strickvl — APPROVED** [review]
LGTM from my side, though agree that giving people simpler choices (or just one choice) + less text to read is probably the way to go.

**Comment by stefannica** [comment]
> @stefannica id rather say replace "easy cloud deployment wizard" and "easy cloud registration wizard" with "in-browser stack deployment wizard" and "stack registration wizard"

@htahir1 done

**Review by htahir1 — APPROVED** [review]
LGTM! Please do backport!

**Comment by stefannica** [comment]
> LGTM! Please do backport!

@htahir1 backported: https://docs.zenml.io/how-to/stack-deployment/deploy-a-cloud-stack-with-terraform

**Pull request merged**